server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -925,17 +941,41 @@ private static Map<String, IndexId> renamedIndices(
         return Collections.unmodifiableMap(renamedIndices);
     }
 
-    private static String renameIndex(String index, RestoreSnapshotRequest request, boolean partOfDataStream) {
-        String renamedIndex = index;
-        if (request.renameReplacement() != null && request.renamePattern() != null) {
-            partOfDataStream = partOfDataStream && index.startsWith(DataStream.BACKING_INDEX_PREFIX);
-            if (partOfDataStream) {
-                index = index.substring(DataStream.BACKING_INDEX_PREFIX.length());
-            }
-            renamedIndex = index.replaceAll(request.renamePattern(), request.renameReplacement());
-            if (partOfDataStream) {
-                renamedIndex = DataStream.BACKING_INDEX_PREFIX + renamedIndex;
-            }
+    private static String renameIndex(String index, RestoreSnapshotRequest request, DataStreamIndices dataStreamIndices) {
+        if (request.renameReplacement() == null || request.renamePattern() == null) {
+            return index;
+        }
+        if (dataStreamIndices.isBackingIndex(index)) {
+            return renameBackingIndex(index, request);
+        } else if (dataStreamIndices.isFailureStore(index)) {
+            return renameFailureIndex(index, request);
+        }
+        return renameIndex(index, request, (String) null);
+    }
+
+    private static String renameBackingIndex(String index, RestoreSnapshotRequest request) {
+        if (request.renameReplacement() == null || request.renamePattern() == null) {
+            return index;
+        }
+        return renameIndex(index, request, DataStream.BACKING_INDEX_PREFIX);
+    }
+
+    private static String renameFailureIndex(String index, RestoreSnapshotRequest request) {
```

Comment:
This looks identical to the above method. I personally think it scans a little better while reading, but the duplication makes me think that it's probably better to merge all of it and control the logic based on which prefix is present (or not present for non-data streams)

I think we're also losing a check here in `renameIndex()` if we're restoring an index that should not be renamed. Not sure from just reviewing if that check is important right now, but it existed beforehand and makes me worry about its removal.

Reply:
I thought I commented this already but I guess I didn't; the if-statements in `renameBackingIndex` and `renameFailureIndex` are already covered by the `renameIndex` above. So, after removing those if-statements, it probably makes even more sense to merge the methods.